ports/stm32/machine_i2s.c

```diff
@@ -0,0 +1,1109 @@
+        { MP_QSTR_bits,     MP_ARG_KW_ONLY | MP_ARG_REQUIRED | MP_ARG_INT,   {.u_int = -1} },
+        { MP_QSTR_format,   MP_ARG_KW_ONLY | MP_ARG_REQUIRED | MP_ARG_INT,   {.u_int = -1} },
+        { MP_QSTR_rate,     MP_ARG_KW_ONLY | MP_ARG_REQUIRED | MP_ARG_INT,   {.u_int = -1} },
+        { MP_QSTR_ibuf,     MP_ARG_KW_ONLY | MP_ARG_REQUIRED | MP_ARG_INT,   {.u_int = -1} },
+    };
+
+    mp_arg_val_t args[MP_ARRAY_SIZE(allowed_args)];
+    mp_arg_parse_all(n_pos_args, pos_args, kw_args, MP_ARRAY_SIZE(allowed_args), allowed_args, args);
+
+    memset(&self->hi2s, 0, sizeof(self->hi2s));
+
+    //
+    // ---- Check validity of arguments ----
+    //
+
+    // are I2S pin assignments valid?
+    const pin_af_obj_t *pin_af;
+
+    // is SCK valid?
+    if (mp_obj_is_type(args[ARG_sck].u_obj, &pin_type)) {
+        pin_af = pin_find_af(MP_OBJ_TO_PTR(args[ARG_sck].u_obj), AF_FN_I2S, self->i2s_id);
+        if (pin_af->type != AF_PIN_TYPE_I2S_CK) {
+            mp_raise_ValueError(MP_ERROR_TEXT("invalid SCK pin"));
+        }
+    } else {
+        mp_raise_ValueError(MP_ERROR_TEXT("SCK not a Pin type"));
+    }
+
+    // is WS valid?
+    if (mp_obj_is_type(args[ARG_ws].u_obj, &pin_type)) {
+        pin_af = pin_find_af(MP_OBJ_TO_PTR(args[ARG_ws].u_obj), AF_FN_I2S, self->i2s_id);
+        if (pin_af->type != AF_PIN_TYPE_I2S_WS) {
+            mp_raise_ValueError(MP_ERROR_TEXT("invalid WS pin"));
+        }
+    } else {
+        mp_raise_ValueError(MP_ERROR_TEXT("WS not a Pin type"));
+    }
+
+    // is SD valid?
+    if (mp_obj_is_type(args[ARG_sd].u_obj, &pin_type)) {
+        pin_af = pin_find_af(MP_OBJ_TO_PTR(args[ARG_sd].u_obj), AF_FN_I2S, self->i2s_id);
+        if (pin_af->type != AF_PIN_TYPE_I2S_SD) {
+            mp_raise_ValueError(MP_ERROR_TEXT("invalid SD pin"));
+        }
+    } else {
+        mp_raise_ValueError(MP_ERROR_TEXT("SD not a Pin type"));
+    }
+
+    // is Mode valid?
+    uint16_t i2s_mode = args[ARG_mode].u_int;
+    if ((i2s_mode != (I2S_MODE_MASTER_RX)) &&
+        (i2s_mode != (I2S_MODE_MASTER_TX))) {
+        mp_raise_ValueError(MP_ERROR_TEXT("invalid mode"));
+    }
+
+    // is Bits valid?
+    int8_t i2s_bits = args[ARG_bits].u_int;
+    if ((i2s_bits != 16) &&
+        (i2s_bits != 32)) {
+        mp_raise_ValueError(MP_ERROR_TEXT("invalid bits"));
+    }
+
+    // is Format valid?
+    format_t i2s_format = args[ARG_format].u_int;
+    if ((i2s_format != MONO) &&
+        (i2s_format != STEREO)) {
+        mp_raise_ValueError(MP_ERROR_TEXT("invalid format"));
+    }
+
+    // is Rate valid?
+    // Not checked
+
+    // is Ibuf valid?
+    int32_t ring_buffer_len = args[ARG_ibuf].u_int;
+    if (ring_buffer_len > 0) {
+        uint8_t *buffer = m_new(uint8_t, ring_buffer_len);
+        self->ring_buffer_storage = buffer;
+        ringbuf_init(&self->ring_buffer, buffer, ring_buffer_len);
+    } else {
+        mp_raise_ValueError(MP_ERROR_TEXT("invalid buf"));
```

Comment:
"ibuf"

Reply:
fixed in https://github.com/miketeachman/micropython/commit/0f7b59c92e8173d86f596097f9f8fc5ef3e3cfd5